onchain/rollups/contracts/consensus/AbstractConsensus.sol

```diff
@@ -4,12 +4,47 @@
 pragma solidity ^0.8.8;
 
 import {IConsensus} from "./IConsensus.sol";
+import {InputRange} from "../common/InputRange.sol";
+import {LibInputRange} from "../library/LibInputRange.sol";
 
-/// @title Abstract Consensus
-/// @notice An abstract contract that partially implements `IConsensus`.
+/// @notice Stores epoch hashes for several DApps and input ranges.
+/// @dev This contract was designed to be inherited by implementations of the `IConsensus` interface
+/// that only need a simple mechanism of storage and retrieval of epoch hashes.
 abstract contract AbstractConsensus is IConsensus {
-    /// @notice Emits an `ApplicationJoined` event with the message sender.
-    function join() external override {
-        emit ApplicationJoined(msg.sender);
+    using LibInputRange for InputRange;
+
+    /// @notice Indexes epoch hashes by DApp address, first input index and last input index.
+    mapping(address => mapping(uint256 => mapping(uint256 => bytes32)))
+        private _epochHashes;
+
+    /// @notice Get the epoch hash for a certain DApp and input range.
+    /// @param dapp The DApp contract address
+    /// @param r The input range
+    /// @return epochHash The epoch hash
+    /// @dev For claimed epochs, returns the epoch hash of the last accepted claim.
+    /// @dev For unclaimed epochs, returns `bytes32(0)`.
+    function getEpochHash(
+        address dapp,
+        InputRange calldata r
+    ) public view override returns (bytes32 epochHash) {
+        epochHash = _epochHashes[dapp][r.firstInputIndex][r.lastInputIndex];
+    }
+
+    /// @notice Accept a claim.
+    /// @param dapp The DApp contract address
+    /// @param r The input range
+    /// @param epochHash The epoch hash
+    /// @dev Raises an `InputRangeEmpty` error if `r` is empty.
```

Comment:
There is no possibility of `r` not being assigned to a value. It is a parameter, which can be evaluated into a value. Of course, in Solidity, variables which are not assigned to some value explicitly are implicitly assigned to the default value of the variable type... But function arguments must be provided explicitly. Solidity doesn't support default values for parameters.

Meanwhile, the discussion of whether "empty" might be ambiguous is completely valid. Maybe if we phrase it like `represents the empty set` rather than `is empty` would make it clearer? The Wikipedia article on [intervals](https://en.wikipedia.org/wiki/Interval_(mathematics)) shows this is a common phrase:

> Each interval (a, a), [a, a), and (a, a] represents the [empty set](https://en.wikipedia.org/wiki/Empty_set), whereas [a, a] denotes the singleton set {a}. When a > b, all four notations are usually taken to represent the empty set.

Reply:
This has been addressed by commit 6801e97.

Reply:
> There is no possibility of r not being assigned to a value

Wouldn't it be possible to declare `r` first without assignment and pass it via function call?

Reply:
I think "invalid" is shorter and simpler than "represents the empty set". 
"empty set" sounds like a term that some of the readers may Wikipedia it, but "invalid" is straightforward.

Reply:
I find the term "invalid" too vague.
What makes an input range invalid?
That's why I prefer to be explicit, and say "represents the empty set".

Reply:
I don't think the concept of an empty set is alien to the average `rollups-contract` user. There are far more complex concepts in our code base.

Reply:
> Wouldn't it be possible to declare `r` first without assignment and pass it via function call?

Yes, but then you're making the decision to pass an unassigned variable to a function call, which is weird.